File: internal/seed/seed.go
package seed

import (
	"database/sql"

	"github.com/xebia/ai-assisted-se-claude-exercise/internal/store"
)

// Run populates the database with sample data.
// It is idempotent only in the sense that it always inserts fresh rows;
// call it only when the database is empty or when a reseed is explicitly requested.
func Run(db *sql.DB) error {
	authors := store.NewAuthorStore(db)
	books := store.NewBookStore(db)
	reviews := store.NewReviewStore(db)

	type authorSeed struct{ name, bio string }
	authorSeeds := []authorSeed{
		{"George Orwell", "English novelist and essayist."},
		{"Ursula K. Le Guin", "American author of speculative fiction."},
		{"Cormac McCarthy", "American novelist and playwright."},
		{"Toni Morrison", "American novelist and Nobel laureate."},
		{"Haruki Murakami", "Japanese author of surrealist fiction."},
	}

	authorIDs := make([]int, len(authorSeeds))
	for i, a := range authorSeeds {
		author, err := authors.Create(a.name, a.bio)
		if err != nil {
			return err
		}
		authorIDs[i] = author.ID
	}

	type bookSeed struct {
		title     string
		isbn      string
		year      int
		authorIdx int
	}
	bookSeeds := []bookSeed{
		{"1984", "978-0451524935", 1949, 0},
		{"Animal Farm", "978-0451526342", 1945, 0},
		{"Homage to Catalonia", "978-0156421171", 1938, 0},
		{"The Left Hand of Darkness", "978-0441478125", 1969, 1},
		{"The Dispossessed", "978-0061054884", 1974, 1},
		{"A Wizard of Earthsea", "978-0547773742", 1968, 1},
		{"Blood Meridian", "978-0679728757", 1985, 2},
		{"No Country for Old Men", "978-0307387899", 2005, 2},
		{"The Road", "978-0307387899", 2006, 2},
		{"Beloved", "978-1400033416", 1987, 3},
		{"Song of Solomon", "978-1400033423", 1977, 3},
		{"Sula", "978-1400033430", 1973, 3},
		{"Norwegian Wood", "978-0375704024", 1987, 4},
		{"Kafka on the Shore", "978-1400079278", 2002, 4},
		{"The Wind-Up Bird Chronicle", "978-0679775430", 1994, 4},
	}

	bookIDs := make([]int, len(bookSeeds))
	for i, b := range bookSeeds {
		book, err := books.Create(b.title, authorIDs[b.authorIdx], b.isbn, b.year)
		if err != nil {
			return err
		}
		bookIDs[i] = book.ID
	}

	type reviewSeed struct {
		bookIdx    int
		rating     int
		reviewText string
	}
	reviewSeeds := []reviewSeed{
		{0, 5, "A chilling vision of totalitarianism."},
		{1, 4, "A sharp political allegory."},
		{3, 5, "A masterpiece of speculative fiction."},
		{6, 4, "Brutal and poetic."},
		{8, 5, "Haunting and unforgettable."},
		{9, 5, "Morrison at her very best."},
		{12, 4, "Melancholic and beautifully written."},
		{13, 3, "Imaginative but occasionally meandering."},
		{14, 5, "Surreal and utterly gripping."},
		{4, 4, "A profound meditation on freedom."},
	}

	for _, rv := range reviewSeeds {
		if _, err := reviews.Create(bookIDs[rv.bookIdx], rv.rating, rv.reviewText); err != nil {
			return err
		}
	}
	return nil
}

// IsEmpty reports whether the authors table has no rows.
func IsEmpty(db *sql.DB) (bool, error) {
	var n int
	err := db.QueryRow(`SELECT COUNT(*) FROM authors`).Scan(&n)
	return n == 0, err
}

File: bookstore-ts/src/seed/seed.ts
import type { Database } from "bun:sqlite";
import { AuthorStore } from "../store/author";
import { BookStore } from "../store/book";
import { ReviewStore } from "../store/review";

export function run(db: Database): void {
  const authors = new AuthorStore(db);
  const books = new BookStore(db);
  const reviews = new ReviewStore(db);

  const authorSeeds: { name: string; bio: string }[] = [
    { name: "George Orwell", bio: "English novelist and essayist." },
    { name: "Ursula K. Le Guin", bio: "American author of speculative fiction." },
    { name: "Cormac McCarthy", bio: "American novelist and playwright." },
    { name: "Toni Morrison", bio: "American novelist and Nobel laureate." },
    { name: "Haruki Murakami", bio: "Japanese author of surrealist fiction." },
  ];

  const authorIDs = authorSeeds.map((a) => authors.create(a.name, a.bio).id);

  const bookSeeds: { title: string; isbn: string; year: number; authorIdx: number }[] = [
    { title: "1984", isbn: "978-0451524935", year: 1949, authorIdx: 0 },
    { title: "Animal Farm", isbn: "978-0451526342", year: 1945, authorIdx: 0 },
    { title: "Homage to Catalonia", isbn: "978-0156421171", year: 1938, authorIdx: 0 },
    { title: "The Left Hand of Darkness", isbn: "978-0441478125", year: 1969, authorIdx: 1 },
    { title: "The Dispossessed", isbn: "978-0061054884", year: 1974, authorIdx: 1 },
    { title: "A Wizard of Earthsea", isbn: "978-0547773742", year: 1968, authorIdx: 1 },
    { title: "Blood Meridian", isbn: "978-0679728757", year: 1985, authorIdx: 2 },
    { title: "No Country for Old Men", isbn: "978-0307387899", year: 2005, authorIdx: 2 },
    { title: "The Road", isbn: "978-0307387899", year: 2006, authorIdx: 2 },
    { title: "Beloved", isbn: "978-1400033416", year: 1987, authorIdx: 3 },
    { title: "Song of Solomon", isbn: "978-1400033423", year: 1977, authorIdx: 3 },
    { title: "Sula", isbn: "978-1400033430", year: 1973, authorIdx: 3 },
    { title: "Norwegian Wood", isbn: "978-0375704024", year: 1987, authorIdx: 4 },
    { title: "Kafka on the Shore", isbn: "978-1400079278", year: 2002, authorIdx: 4 },
    { title: "The Wind-Up Bird Chronicle", isbn: "978-0679775430", year: 1994, authorIdx: 4 },
  ];

  const bookIDs = bookSeeds.map(
    (b) => books.create(b.title, authorIDs[b.authorIdx]!, b.isbn, b.year).id,
  );

  const reviewSeeds: { bookIdx: number; rating: number; reviewText: string }[] = [
    { bookIdx: 0, rating: 5, reviewText: "A chilling vision of totalitarianism." },
    { bookIdx: 1, rating: 4, reviewText: "A sharp political allegory." },
    { bookIdx: 3, rating: 5, reviewText: "A masterpiece of speculative fiction." },
    { bookIdx: 6, rating: 4, reviewText: "Brutal and poetic." },
    { bookIdx: 8, rating: 5, reviewText: "Haunting and unforgettable." },
    { bookIdx: 9, rating: 5, reviewText: "Morrison at her very best." },
    { bookIdx: 12, rating: 4, reviewText: "Melancholic and beautifully written." },
    { bookIdx: 13, rating: 3, reviewText: "Imaginative but occasionally meandering." },
    { bookIdx: 14, rating: 5, reviewText: "Surreal and utterly gripping." },
    { bookIdx: 4, rating: 4, reviewText: "A profound meditation on freedom." },
  ];

  for (const rv of reviewSeeds) {
    reviews.create(bookIDs[rv.bookIdx]!, rv.rating, rv.reviewText);
  }
}

export function isEmpty(db: Database): boolean {
  const row = db.query(`SELECT COUNT(*) AS n FROM authors`).get() as { n: number };
  return row.n === 0;
}
